lib/dangermattic/plugins/labels_checker.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module Danger
+  # Plugin to check for labels in a PR.
+  class LabelsChecker < Plugin
+    DEFAULT_DO_NOT_MERGE_LABEL = 'Do Not Merge'
+    DEFAULT_LABELS_WARNING = 'Please add the required labels for this project.'
+    DEFAULT_REQUIRED_LABELS = [/.*\S.*/].freeze
+
+    # Checks if a PR is missing labels or is marked as 'do not merge'.
+    #
+    # @param do_not_merge_label [String] The label indicating that a merge should not be performed.
+    #   Defaults to DEFAULT_DO_NOT_MERGE_LABEL if not provided.
+    # @param required_labels [Array<String>] The list of labels that must be present for a merge operation to proceed.
+    #   Defaults to an empty array if not provided.
+    # @param required_labels_warning [String] The warning message displayed if the required labels are not present.
+    #   Defaults to DEFAULT_LABELS_WARNING if not provided.
+    #
+    # @return [void]
+    def check(do_not_merge_label: DEFAULT_DO_NOT_MERGE_LABEL, required_labels: DEFAULT_REQUIRED_LABELS,
+              required_labels_warning: DEFAULT_LABELS_WARNING)
+      github_labels = danger.github.pr_labels
+
+      pr_has_all_required_labels = github_labels.all? do |pr_label|
+        required_labels.any? do |required_label|
+          pr_label =~ required_label
+        end
+      end
```

Comment:
Would be nice to capture the list of missing required labels, so we can print it in the error message
```suggestion
      missing_required_labels = required_labels.reject do |required_label|
        github_labels.any? { |pr_label| pr_label =~ required_label }
      end
```

Reply:
Also, it would be nice for the Plugin to support the elements of `required_labels` to be a either a `RegExp` or just a plain `String` (and in such case, we'd check for an exact match instead of `~=`). Probably a rare use case to require a specific and fixed label on all PRs tho 🤔

Reply:
Another use case I just thought of… but that can probably wait for the second iteration of this Plugin, is to conditionalize the required labels based on the base branch.

For example, if the PR targets the branch `release/*` (which corresponds to the code-freeze in most of our repos, similar to `beta` branch in Tumblr), then require either the `bug` or `bugfix-exemption` label, otherwise not. Maybe using an API like this?
```ruby
labels_checker.check(required_labels: {
  /^release\/.*/: /^(bug|bugfix-exemption)$/,
  'develop': /^feature: .*/
})
```
Though maybe it would be simpler to let the call site / client repo do the test in their `Dangerfile` themselves, like this?
```ruby
if github.branch_for_base.start_with?('release/')
  labels_checker.check(required_labels: [/^(bug|bugfix-exemption)$/])
else
  labels_checker.check(required_labels: [/^feature:/])
end
```
Probably a matter of balance to find between how much we want to be within the Plugin (to avoid each client repo to repeat the same kind of code pattern over and over) vs how much we want to be written in the `Dangerfile` (which could allow more flexibility, but also might lead to code repetition between `Dangerfiles` and kinda defeats part of the benefits of DRY-ing code into a Plugin)

I guess we'll see as we start adopting this plugin in repos how we should make that plugin evolve, tho, so maybe better to wait and see concrete use cases first?

Reply:
> Probably a matter of balance to find between how much we want to be within the Plugin (to avoid each client repo to repeat the same kind of code pattern over and over) vs how much we want to be written in the Dangerfile

Exactly. My intuition for this specific case is that the plugins should fulfil the general case as much as possible, with this kind of check done outside of the plugin and avoiding conditional logic based on external factors -- the conditional check could also be part of a common `Dangerfile` that repos can include, if it applies to enough applications.

But as you said, I think we need to start seeing once we start adopting it in the repos to have a better idea.